src/rust/catloop/mod.rs

```diff
@@ -547,38 +606,84 @@ impl CatloopLibOS {
         Err(Fail::new(libc::EINVAL, &cause))
     }
 
+    /// Harvests resources on close.
+    fn harvest_resources_on_close(&mut self, qd: QDesc) -> Result<(), Fail> {
+        // If this was an async_close(), harvest resources.
+        let mut qtable: RefMut<IoQueueTable<CatloopQueue>> = self.qtable.borrow_mut();
+        match qtable.get_mut(&qd) {
+            Some(queue) => {
+                // Rollback the port allocation.
+                if let Socket::Active(Some(addr)) | Socket::Passive(addr) = queue.get_socket() {
+                    if EphemeralPorts::is_private(addr.port()) {
+                        if self.ephemeral_ports.borrow_mut().free(addr.port()).is_err() {
+                            // We fail if and only if we attempted to free a port that was not allocated.
+                            // This is unexpected, but if it happens, issue a warning and keep going,
+                            // otherwise we would leave the queue in a dangling state.
+                            warn!("pakc_result(): leaking ephemeral port (port={})", addr.port());
+                        }
+                    }
+                }
+                queue.cancel_pending_ops(Fail::new(libc::ECANCELED, "this queue was closed"));
+                qtable.free(&qd);
+                Ok(())
+            },
+            None => {
+                let cause: String = format!("invalid queue descriptor (qd={:?})", qd);
+                error!("close(): {}", &cause);
+                Err(Fail::new(libc::EBADF, &cause))
+            },
+        }
+    }
+
     /// Constructs an operation result from a scheduler handler and queue token pair.
     pub fn pack_result(&mut self, handle: TaskHandle, qt: QToken) -> Result<demi_qresult_t, Fail> {
         // Check if the queue token came from the Catloop LibOS.
-        if let Some((ref opcode, _)) = self.catloop_qts.remove(&qt) {
+        if let Some((opcode, catloop_qd)) = self.catloop_qts.remove(&qt) {
             // Check if the queue token concerns an expected operation.
-            if opcode != &demi_opcode_t::DEMI_OPC_ACCEPT && opcode != &demi_opcode_t::DEMI_OPC_CONNECT {
-                let cause: String = format!("unexpected queue token (qt={:?})", qt);
-                error!("pack_result(): {:?}", &cause);
-                return Err(Fail::new(libc::EINVAL, &cause));
+            match opcode {
+                demi_opcode_t::DEMI_OPC_ACCEPT | demi_opcode_t::DEMI_OPC_CONNECT | demi_opcode_t::DEMI_OPC_CLOSE => {},
+                _ => {
+                    let cause: String = format!("unexpected queue token (qt={:?})", qt);
+                    error!("pack_result(): {:?}", &cause);
+                    return Err(Fail::new(libc::EINVAL, &cause));
+                },
+            }
+
+            // If this was an async_close(), harvest resources.
```

Comment:
I don't think this clean up code is getting called because the qt would be in the list of catmem qts for a close on the normal path.

Reply:
This is more a sanity check, as we do for accept and connect, right?